ethereum/contracts/Decoder.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+library Decoder {
+
+    // Decoes a SCALE encoded uint256
+    function decodeUint256(bytes memory data) internal pure returns (uint256) {
+        bytes memory reversed = reverseBytes(data);
+        uint256 lEndian = bytesToLittleEndian(reversed);
+        return lEndian;
+    }
+
+    // Convert bytes (big endian) to little endian format
+    function bytesToLittleEndian(bytes memory arr) internal pure returns (uint256) {
+        uint256 number;
+        for(uint i = 0; i < arr.length; i++){
+            number = number + uint(uint8(arr[i])) * (2 ** (8 * (arr.length - (i + 1))));
+        }
+        return number;
+    }
+
+     // Reverse an array of bytes in place
+    function reverseBytes(bytes memory arr) internal pure returns (bytes memory)
+    {
+        for (uint i = 0; i < arr.length/2; i++) {
+            bytes1 current = arr[i];
+            uint256 otherIndex = arr.length - i - 1;
+            arr[i] = arr[otherIndex];
+            arr[otherIndex] = current;
+        }
+        return arr;
+    }
+
+    // Slice a section from a byte array.
+    // Inspired by github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol.
+    function slice(bytes memory _bytes, uint256 _start, uint256 _length) internal pure returns (bytes memory) {
```

Comment:
What is the reason for implementing such a low level function? Why not just use the slice literal? https://ethereum.stackexchange.com/questions/78559/how-can-i-slice-bytes-strings-and-arrays-in-solidity 

If we want to keep this logic, could you also copy all the comments from github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol for context? Thanks!

Reply:
Because index range access is only supported for dynamic calldata arrays and copying calldata to memory is not implemented in the version of Solidity we're using. Once copying calldata to memory is updated we'll be able to use the slice literal. Open to suggestions on this one, this was primarily just a workable solution for M4.

Related: https://github.com/ethereum/solidity/issues/7929, https://github.com/ethereum/solidity/issues/8210

Reply:
Okay, could you add that comment in code? Thanks!